crmsh/join_lock.py

```diff
@@ -0,0 +1,139 @@
+    DETECT_INTERVAL = 15
+    UPDATE_INTERVAL = 10
+
+    def __init__(self, target_node):
+        """
+        Init function
+        """
+        self.thread_stop_event = None
+        self.target_node = target_node
+
+    def _run(self, cmd):
+        """
+        Wrapper utils.run_cmd_on_remote, exit when meeting error
+        """
+        rc, out, err = utils.run_cmd_on_remote(cmd, self.target_node)
+        if rc != 0 and err:
+            bootstrap.error(err)
+        return rc, out
+
+    def _detect_lock_file(self):
+        """
+        Detect whether target node has the lock file
+        """
+        rc, _ = self._run(self.DETECT_CMD)
+        return rc == 0
+
+    def _get_last_timestamp(self):
+        """
+        Tail the last line in lock file
+        """
+        rc, out = self._run(self.TAIL_CMD)
+        if rc == 0 and out:
+            try:
+                current_node, timestamp = out.split()
+            except ValueError:
+                bootstrap.error("Invalid format for {}(should be \"node timestamp\")".format(self.JOIN_LOCK_FILE))
+            return current_node, timestamp
+
+    def _remove_lock_file(self):
+        """
+        Remove lock file
+        """
+        self._run(self.RM_CMD)
+
+    def _touch_lock_file(self):
+        """
+        Touch lock file
+        """
+        self._run(self.TOUCH_CMD)
+
+    def _wait(self):
+        """
+        Wait 15 seconds while detect target node has lock file
+        """
+        warned_once = False
+        origin_timestamp = None
+
+        while self._detect_lock_file():
+
+            current_node, timestamp = self._get_last_timestamp()
+
+            if origin_timestamp:
+                res = int(timestamp) - int(origin_timestamp)
+                # current join node might has exceptions
+                if res < self.UPDATE_INTERVAL:
+                    # remove lock file on target node
+                    self._remove_lock_file()
+                    break
+            else:
+                origin_timestamp = timestamp
+
+            if not warned_once:
+                warned_once = True
+                bootstrap.warn("Node {} still joining, wait...".format(current_node))
+
+            time.sleep(self.DETECT_INTERVAL)
+
+    def _update_timestamp(self):
+        """
+        Update timestamp(10s interval) in lock file on target node, format is:
+            nodename timestamp
+        """
+        while not self.thread_stop_event.is_set():
+            # TODO utils.run_cmd_on_remote have issue to run in thread
+            if not bootstrap.invoke("ssh root@{} \"{}\"".format(self.target_node, self.UPDATE_CMD)):
+                bootstrap.error("Can't invoke {} on {}".format(self.UPDATE_CMD, self.target_node))
+            time.sleep(self.UPDATE_INTERVAL)
+
+    @contextmanager
+    def lock(self):
+        """
+        Create lock file on target node
+        """
+        # wait if other node was joining
+        self._wait()
```

Comment:
We need to change some logic here. The `touch` (or maybe other system) must be the 1st thing we do. Otherwise, if 2 nodes start the lock in the same moment, both will pass and get the log.
We need to have something atomic.
Maybe we need to combine the `touch` with a `mkdir`. This second will fail if the folder is there. The touch is quite risky, as it works when the file doesn't exists and when it exists. And it is used after the wait.